### ISSUE TITLE
Custom Lambda Authorizers not working with HTTPApi

### ISSUE THREAD

**Issue by mkamioner** [aws/serverless-application-model#2049]
### Description:

When creating a `AWS::Serverless::HttpApi` with a custom lambda authorizer, the API fails to be invoked with a 500 error.

### Steps to reproduce:

Here is an example of the template I was using to reproduce the bug
```yaml
AWSTemplateFormatVersion: 2010-09-09
Transform: AWS::Serverless-2016-10-31

Resources:
  SomeHttpApi:
    Type: AWS::Serverless::HttpApi
    Properties:
      CorsConfiguration:
        AllowHeaders: ["*"]
        AllowMethods: ["*"]
        AllowOrigins: ["*"]
      DefaultRouteSettings:
        DetailedMetricsEnabled: true
      Auth:
        DefaultAuthorizer: MyCustomAuthorizer
        Authorizers:
          MyCustomAuthorizer:
            AuthorizerPayloadFormatVersion: "2.0"
            EnableSimpleResponses: true
            FunctionArn: !GetAtt SomeAuthorizationFunction.Arn
            FunctionInvokeRole: !GetAtt SomeAuthorizationFunctionRole.Arn

  SomeBasicApi:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: some-basic-api
      Handler: index.handler
      Runtime: python3.8
      InlineCode: |
        def handler(event, context):
          return {"body": "Hello, world!"}
      MemorySize: 128
      Events:
        Api:
          Type: HttpApi
          Properties:
            Method: GET
            Path: /hello
            ApiId: !Ref SomeHttpApi
      Policies:
        - AWSLambdaBasicExecutionRole

  SomeAuthorizationFunction:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: some-authorization-function
      Handler: index.handler
      Runtime: python3.8
      InlineCode: |
        def handler(event, context):
            return {
              "isAuthorized": True,
              "context": {
                "exampleKey": "exampleValue"
              }
            }
      MemorySize: 128
      Policies:
        - AWSLambdaBasicExecutionRole
```

### Observed result:

When calling the API without `Auth` property, everything worked as expected. Once added the HTTP call fails with a 500 error. The authorizer lambda is not invoked. The logs of the API do not reveal anything wrong...

### Expected result:

I would expect the authorize lambda to execute properly and continue to allow the user access to the API

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: Mac
2. If using SAM CLI, `sam --version`:
3. AWS region: us-east-1

`Add --debug flag to any SAM CLI commands you are running`

**Comment by c2tarun** [comment]
On some initial investigation this feels like a bug in SAM.
### TLDR;
ApiGateway do not have permission to invoke Authorizer Lambda function. Temporary fix to unblock you:
Create custom role for your authorizer function and allow API Gateway to assume that function and add Lambda invocation permission to that role as well :( (I know it sounds like a hack).

Here is how diff looks like:

```
❯ diff -u issue-template.yaml template.yaml
--- issue-template.yaml	2021-06-03 10:01:29.000000000 -0700
+++ template.yaml	2021-06-03 10:00:50.000000000 -0700
@@ -1,4 +1,4 @@
-AWSTemplateFormatVersion: 2010-09-09
+AWSTemplateFormatVersion: "2010-09-09"
 Transform: AWS::Serverless-2016-10-31

 Resources:
@@ -43,6 +43,7 @@
   SomeAuthorizationFunction:
     Type: AWS::Serverless::Function
     Properties:
+      Role: !GetAtt SomeAuthorizationFunctionRole.Arn
       FunctionName: some-authorization-function
       Handler: index.handler
       Runtime: python3.8
@@ -57,3 +58,41 @@
       MemorySize: 128
       Policies:
         - AWSLambdaBasicExecutionRole
+
+  SomeAuthorizationFunctionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Action:
+              - sts:AssumeRole
+            Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+          - Action:
+              - sts:AssumeRole
+            Effect: Allow
+            Principal:
+              Service:
+                - apigateway.amazonaws.com
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+      Tags:
+        - Key: lambda:createdBy
+          Value: SAM
+
+  LambdaInvocationPolicy:
+    Type: AWS::IAM::Policy
+    Properties:
+      PolicyName: LambdaInvocationPolicy
+      PolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Action: lambda:InvokeFunction
+            Resource: !GetAtt SomeAuthorizationFunction.Arn
+      Roles:
+        - !Ref SomeAuthorizationFunctionRole
+

```

### Detailed explanation:
When SAM transforms HttpApi event provided in SAM function it generates an `AWS::ApiGatewayV2::Api` type resource and configure it as per [ApiGateway OpenApi guidelines](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions.html). This results in a resource similar to below:

```
  SomeHttpApi:
    Type: AWS::ApiGatewayV2::Api
    Properties:
      Body:
        openapi: 3.0.1
        info:
          version: "1.0"
          title:
            Ref: AWS::StackName
        paths:
          /hello:
            get:
              x-amazon-apigateway-integration:
                 . . .
        x-amazon-apigateway-cors:
          . . .
        components:
          securitySchemes:
            MyCustomAuthorizer:
              type: apiKey
              name: Unused
              in: header
              x-amazon-apigateway-authorizer:
                type: request
                authorizerUri:
                  Fn::Sub:
                    - arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${__FunctionArn__}/invocations
                    - __FunctionArn__:
                        Fn::GetAtt:
                          - SomeAuthorizationFunction
                          - Arn
                authorizerCredentials:
                  Fn::GetAtt:
                    - SomeAuthorizationFunctionRole
                    - Arn
                authorizerPayloadFormatVersion: "2.0"
                enableSimpleResponses: true
        tags:
            . . .
```
Notice field `authorizerCredentials` is simply using the Role for Authorizer lambda function. Based on [authorizerCredentials definition here](https://docs.aws.amazon.com/apigateway/api-reference/resource/authorizer/#authorizerCredentials) it should be a role which have permission to invoke the authorizer. The problem with using role for Lambda function is:
1. ApiGateway do not have permission to assume this role.
2. Role itself do not have permission to invoke the function.

A proper fix would be to generate a new Authorizer role and use that instead of function role. However, this means that the functionality have been broken for a while which doesn't sound correct. I'll sync more with authors of this feature and add more details to this issue.

**Comment by c2tarun** [comment]
Ignore my previous comment. Issue here is your FunctionInvokeRole is pointing to a role which do not have permission to invoke the Authorizer function. That is why API Gateway is failing to invoke that function. Create another role that can be assumed by APIGateway and grant that role invocation permission:

```
  SomeAuthorizationInvokationRole:
     Type: AWS::IAM::Role
     Properties:
       AssumeRolePolicyDocument:
         Version: "2012-10-17"
         Statement:
           - Action:
               - sts:AssumeRole
             Effect: Allow
             Principal:
               Service:
                 - apigateway.amazonaws.com
       Tags:
         - Key: lambda:createdBy
           Value: SAM
  LambdaInvocationPolicy:
    Type: AWS::IAM::Policy
    Properties:
      PolicyName: LambdaInvocationPolicy
      PolicyDocument:
        Version: "2012-10-17"
        Statement:
          - Effect: Allow
            Action: lambda:InvokeFunction
            Resource: !GetAtt SomeAuthorizationFunction.Arn
      Roles:
        - !Ref SomeAuthorizationInvokationRole
 ```

Use this role as your `FunctionInvokeRole` and this will work.

**Comment by tjvc** [comment]
@c2tarun thanks for the solution you  provided above - that fixed the same issue for me. I think it might also be possible to achieve the same with a resource-based policy?

```yaml
AuthorizerPermission:
  Type: AWS::Lambda::Permission
  Properties:
    Action: lambda:InvokeFunction
    FunctionName: !GetAtt AuthorizationFunction.Arn
    Principal: apigateway.amazonaws.com
    # https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html
    SourceArn: !Sub arn:aws:execute-api:${AWS::Region}:${AWS::AccountId}:${api}/authorizers/*
```

I'm wondering though if your first comment was also correct. This does seem like a bug, particularly based on the documentation, which does not mention the need to grant permissions to invoke the authorizer function: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-controlling-access-to-apis-lambda-authorizer.html.

Should this issue be re-opened, or at least the docs updated to refer to the need to explicitly grant permissions to the gateway to invoke the authorizer function?
